src/optimizer/plugins/SimplifyArrayIndexingOptimizer.ts

```diff
@@ -0,0 +1,123 @@
+import { fshrules, fhirtypes, utils } from 'fsh-sushi';
+import { OptimizerPlugin } from '../OptimizerPlugin';
+import { Package } from '../../processor';
+import { ExportableCaretValueRule } from '../../exportable';
+
+const singletonArrayElements: Map<string, fhirtypes.PathPart[]> = new Map();
+
+export default {
+  name: 'simplify_array_indexing',
+  description: 'Replace numeric indices with soft indexing',
+  optimize(pkg: Package): void {
```

Comment:
I think it might be necessary to declare a `runsAfter` for this extension. Specifically, I think it needs to run after `CombineCodingAndQuantityValuesOptimizer`. That optimizer relies on rule path equality by simple string comparison, and does not attempt to perform path calculations that involve soft indices. So, if the soft indexing were applied before that optimizer, it would be unable to combine any rules.

Thinking about this scenario more generally, this optimizer should probably be run near the end. Otherwise, we'd need to change any optimizer that modifies the list of rules to be able to handle path comparisons involving soft indices. Fortunately, we've done a good job with naming our optimizers, so it looks like this would need to run after the combiners:
- combine_card_and_flag_rules
- combine_coding_and_quantity_values
- combine_contains_rules
and removers:
- remove_choice_slicing_rules
- remove_default_extension_context_rules
- remove_extension_slicing_rules
- remove_extension_url_assignment_rules
- remove_generated_text_rules
- remove_implied_zero_zero_card_rules
- remove_publisher_derived_date_rules

I'm not 100% that every optimizer in those lists can interact unfavorably with soft indices, but it seems like a safe bet to include all of those as preconditions for applying soft indexing.

Reply:
The problem w/ `runBefore` and `runAfter` is that if we want something to run last, then we need to try to keep up as things get added.  @mint-thompson, @ngfreiter, @julianxcarter -- what do you think about adding a `runLast` boolean?  The only tricky thing is if there are multiple things labeled `runLast` -- but in that case the tie could be broken using `runBefore` and `runAfter`.  The only other thing I can think of would be supporting regex on `runBefore` and `runAfter`, so you could do something like: `runAfter: /^combine/` or even `runAfter: /.*/` (although that still could lead to ties).

Reply:
I like the regex approach because it should be able to work well with the existing ordering strategy we have implemented. Ties are fine as long as they are between optimizers that don't need a strict ordering. While `runLast` is convenient for this case, I think if we were to take that approach, I'd prefer to take it one step further and define subgroups for strict ordering, with partial ordering happening within each subgroup. But, maybe just two subgroups would be enough? In that case, the boolean makes sense.

I feel like it comes down to a choice of one level of sorting or more than one level of sorting, and my preference is to keep it at one level. Subgroups can naturally emerge from having just one level of sorting without having to explicitly define a subgroup.

Reply:
@cmoesel I really like the idea of supporting regex in `runBefore` and `runAfter`. That sounds like it could be pretty useful in other optimizers down the line, and it'll also allow us to effectively have a `runLast` option. If we don't build in an explicit `runLast` option though, then breaking ties could be more difficult. I'd be in favor of adding both.

Reply:
Since you both like the approach of using regular expressions, let's go ahead and give that a whirl.  I think I agree with @mint-thompson that we want to try to keep this at one level -- so let's not add `runLast` just yet; let's see how far we can get w/ the regular expressions. It may be that they're enough for any actual use cases we'll ever have.

Reply:
Didn't see your comment @cmoesel until after I posted mine. Regexs work for me too.

Reply:
Ah, OK.  I actually think that implementation of regex should be pretty simple.  It might just be changing the types like:
```
runBefore?: string[];
```
to:
```
runBefore?: (string|RegExp)[];
```

And then changing where we do the comparison (example below) to either do equality (for string) or test (for RegEx):
```
if (optimizers.some(o => o.name === dependsOn)) {
```

I think it might be even easier than adding a third field (`runLast`) to the mix.

Reply:
Added regex support in [`77ddd51`](https://github.com/FHIR/GoFSH/pull/106/commits/77ddd51820580042db41050654e61133fa74121b).